crates/distribution-types/src/file.rs

```diff
@@ -91,3 +137,87 @@ impl Display for FileLocation {
         }
     }
 }
+
+/// An error that occurs when a `FileLocation` is not a valid URL.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ToUrlError {
+    /// An error that occurs when the base URL in `FileLocation::Relative`
+    /// could not be parsed as a valid URL.
+    InvalidBase {
+        /// The base URL that could not be parsed as a valid URL.
+        base: String,
+        /// The underlying URL parse error.
+        err: url::ParseError,
+    },
+    /// An error that occurs when the base URL could not be joined with
+    /// the relative path in a `FileLocation::Relative`.
+    InvalidJoin {
+        /// The base URL that could not be parsed as a valid URL.
+        base: String,
+        /// The relative path segment.
+        path: String,
+        /// The underlying URL parse error.
+        err: url::ParseError,
+    },
+    /// An error that occurs when the absolute URL in `FileLocation::Absolute`
+    /// could not be parsed as a valid URL.
+    InvalidAbsolute {
+        /// The absolute URL that could not be parsed as a valid URL.
+        absolute: String,
+        /// The underlying URL parse error.
+        err: url::ParseError,
+    },
+    /// An error that occurs when the file path in `FileLocation::Path` is
+    /// not valid UTF-8. We need paths to be valid UTF-8 to be transformed
+    /// into URLs, which must also be UTF-8.
+    PathNotUtf8 {
+        /// The original path that was not valid UTF-8.
+        path: PathBuf,
+    },
+    /// An error that occurs when the file URL created from a file path is not
+    /// a valid URL.
+    InvalidPath {
+        /// The file path URL that could not be parsed as a valid URL.
+        path: String,
+    },
+}
+
+impl std::error::Error for ToUrlError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match *self {
+            ToUrlError::InvalidBase { ref err, .. } => Some(err),
+            ToUrlError::InvalidJoin { ref err, .. } => Some(err),
+            ToUrlError::InvalidAbsolute { ref err, .. } => Some(err),
+            ToUrlError::PathNotUtf8 { .. } => None,
+            ToUrlError::InvalidPath { .. } => None,
+        }
+    }
+}
+
+impl std::fmt::Display for ToUrlError {
```

Comment:
Can we derive those with `thiserror::Error`?

Reply:
I suppose. My hands are very good at writing things out without macros because `thiserror` is IMO not usually worth the dependency. But I suppose that's irrelevant here.